### DISCUSSION TITLE
question about puma architecture

### DISCUSSION THREAD

**Discussion by izenk** [puma/puma#3368] [Q&A] [answered]
Hello.
Have some unclear moments after reading puma architecture docs, exact section:

- Upon startup, Puma listens on a TCP or UNIX socket.
  - The backlog of this socket is configured with a default of 1024, but the actual backlog value is capped by the net.core.somaxconn sysctl value. The backlog determines the size of the queue for unaccepted connections. If the backlog is full, the operating system is not accepting new connections.
  - This socket backlog is distinct from the backlog of work as reported by Puma.stats or the control server. The backlog that Puma.stats refers to represents the number of connections in the process' todo set waiting for a thread from the [ThreadPool](https://github.com/puma/puma/blob/master/lib/puma/thread_pool.rb). - **so, todo set is some kind of internal puma queue where buffered (mean accepted requests) are waiting to be pickup-ed by thread?**  
- By default, a single, separate thread (created by the [Reactor](https://github.com/puma/puma/blob/master/lib/puma/reactor.rb) class) reads and buffers requests from the socket.
  - **(!)** When at least one worker thread is available for work, the reactor thread listens to the socket and accepts a request (if one is waiting). - **does this mean that request from socket will be accepted only if worker thread is available? But If so, in situation when all threads are serving requests (so no threads are available) it means that todo set will always be empty, because no requests will be picked-up from socket backlog, correct? I don't understand how puma stat may show something like "backlog": 5, because it means that requests have been put into todo set, while threads were busy serving requests which contradicts the statement marked with (!) above.**

**Comment by stevenharman** [comment]
> so, todo set is some kind of internal puma queue where buffered (mean accepted requests) are waiting to be pickup-ed by thread?

Yes, that's right. At least, that's true so long as the `queue_requests` option is `true` (which is the default).

As for the question about how/when the "todo" queue is filled from the Socket, I'm also curious about that. Is the "todo" queue bounded somehow? What controls how big it can get?

**Comment by stevenharman** [comment, marked as answer]
I went spelunking to see how the `todo` list (it's an `Array`, not a `Queue`) gets populated. As expected, there's A LOT happening with different threads waiting and signaling for work to happen, or not. Please, someone more familiar correct me where I'm wrong…

From what I can see, the `Puma::Server` (1 per process) will, holds the Socket. During the `Server`'s "run loop" it asks its `Puma::ThreadPool` (pool of worker threads) if the pool is fully busy ([`Puma::ThreadPool#wait_until_not_full`](https://github.com/puma/puma/blob/3169cf607ae3978eabf22c325f02eaefc8ca5c45/lib/puma/thread_pool.rb#L239-L265H)). If it is, the server run loop will block, waiting for the pool to not be fully busy. The idea being, if this Server process's pool of worker threads are all busy, don't bother pulling a new connection off the Socket because some other `Puma::Server` process might not be busy, and so can pull the connection off the socket and handle it.

When one of the worker threads (from the `ThreadPool`) [signals that it's not full](https://github.com/puma/puma/blob/3169cf607ae3978eabf22c325f02eaefc8ca5c45/lib/puma/thread_pool.rb#L139), the server will [accept a connection on the socket, wrap it in a `Puma::Client`, and push it into the `ThreadPool`](https://github.com/puma/puma/blob/3169cf607ae3978eabf22c325f02eaefc8ca5c45/lib/puma/server.rb#L366-L371). Internally, that adds the `Client` to the [`ThreadPool`'s `todo` list, and signals `not_empty`](https://github.com/puma/puma/blob/3169cf607ae3978eabf22c325f02eaefc8ca5c45/lib/puma/thread_pool.rb#L229-L235).

Meanwhile a worker thread (possibly the same one) was [waiting for a `not_empty` signal](https://github.com/puma/puma/blob/3169cf607ae3978eabf22c325f02eaefc8ca5c45/lib/puma/thread_pool.rb#L141). That worker thread wakes up and pops a `Client` off the front of the `todo` list, passing it back to the `Server#process_client` method, which checks the `Client` to see if it's ready to process. If not, it [`#add`s the `Client` to the `Reactor`](https://github.com/puma/puma/blob/3169cf607ae3978eabf22c325f02eaefc8ca5c45/lib/puma/server.rb#L455). This basically ends the worker thread's work loop, so it now signals `not_full` and then `not_empty.wait`s for more things to arrive in the `todo` list.

Adding the `Client` (an I/O wrapper) to the `Reactor` means the `Client` is now being monitored for any data to arrive. When data arrives, the Reactor will (in its own Thread) [wake up and read the data. If all of the data has arrived, it will add the `Client` back to the ThreadPool](https://github.com/puma/puma/blob/3169cf607ae3978eabf22c325f02eaefc8ca5c45/lib/puma/server.rb#L299-L309), meaning back onto the `todo` list. The `Client` is then also [removed from the `Reactor`](https://github.com/puma/puma/blob/3169cf607ae3978eabf22c325f02eaefc8ca5c45/lib/puma/reactor.rb#L119-L120).

Being added back to the `ThreadPool`, the `Client` goes through the same process as above. So that `not_empty.wait` is fulfilled, and a worker thread is back in action. Except this time, when the worker thread pops the `Client` off the `todo` list, and passes it to the `Server#process_client`, the `Client` is ready for processing. So rather than pushing it onto the `Reactor`, the [`Client` is handled (the request is processed)](https://github.com/puma/puma/blob/3169cf607ae3978eabf22c325f02eaefc8ca5c45/lib/puma/server.rb#L467).

So how is it that the `todo` list could get 5 things in it? Timing and concurrent code. That `Server` run loop is trying to accept connections in a loop, and as long as the `ThreadPool` isn't fully busy in any given loop, it will accept a connection, which gets handed over to the `Reactor`. So you could end up with many connections (let's say, 10) in the Reactor. At that point `backlog` is 10, and none of the worker threads are working. But then the Reactor finishes buffering say 5 of those connections. Suddenly 5 of the worker threads are now also busy working those requests. And the backlog is 5.

As soon as one of the workers finishes its work, it will `not_full.signal` and then `not_empty.wait`. This gives the `Server` a chance to unblock its loop and maybe read more off the socket. And thus it goes, round-and-round, with various Threads waking up and waiting, and work being done.

Phew! I'm sure I got some details wrong here. Corrections welcome! Thanks.
